### ISSUE TITLE
Possible Generalization of "pre" for Continuous Input?

### ISSUE THREAD

**Issue by mestinso** [modelica/ModelicaSpecification#3298]
It seems very useful to me if it were possible to generalize the "pre" function for continuous inputs.

For example, I would expect the following behavior for such a generalized function:
For a positive slope: (v-pre(v) > 0) == true
For a negative slope: (pre(v)-v > 0) == true
For a zero slope: (v==pre(v)) == true
For a step change up in signal v: v-pre(v) == the value of the step change at that step change instant

Conceptually pre(v) might expand into something like "v - Modelica.Constants.small*sign(der(v))" for a situation where there is no local discontinuity in order to achieve the above...?

By making this generalization to "pre", continuous versions of falling edge, rising edge, and change blocks should be possible. Related: I would note that Simulink (even with continuous signals) does having falling edge and rising edge blocks which work just fine through step changes whereas detecting a rising edge or a discontinuity can't be done in in a simple/fast/robust way with Modelica. This would bring some parity for those situations.

Additionally, I think this would probably enable a robust/fast implementation to address the following issue:
https://github.com/modelica/ModelicaSpecification/issues/109

Thoughts? Could this be a possible new language feature or are there fundamental issues preventing this?

Alternatively, if a continuous "pre" was too problematic, what about continuous implementation of "edge"?

The following works except it doesn't handle discontinuities in u:
```
block ContinuousEdge
input Real u;
output Real y;
equation
y=der(u)>0;
end ContinousEdge;
```
A robust and general version of the above would effectively behave the same but also work through a step change. For example, a step change up would be similar to a trigger from the sample function.

**Comment by henrikt-ma** [comment]
While I think it would be elegant to generalize `pre` to be a left limit in super-dense time meaning (usual left limit between events and at first round of event iteration, and value from last round of event iteration in subsequent rounds).

However, I'm afraid that implementing this could be challenging due to jumps in `Real` variables between events.  I guess one could try approaching this by making `pre(x)` return the value of `x` at the last accepted step, but I can't say I grasp the consequences of such approximation with regards to things like multi-step integration methods.

Also, would you expect this to work?
```
model JumpAccumulation
  Real x = if noEvent(time < 0.5) then 1.0 else 2.0;
  Real y(start = 0.0, fixed = true) "Accumulated jumps in x";
equation
  der(y) = 0.0;
  when {x > pre(x), x < pre(x)} then
    reinit(y, pre(y) + (x - pre(x)));
  end when;
end JumpAccumulation;
```

**Comment by HansOlsson** [comment]
> 
> However, as I hinted at in various spots in my initial comment, what I really want is the ability to evaluate (rising)edge/fallingEdge/change on continuous signals. In the case of discrete signals, they utilize `pre` to do that, so it got me thinking maybe we need to generalize pre as a precursor to accomplishing edge and change functions. But if it's too problematic, what about just directly creating special edge/change functions for continuous signals that work robustly through discontinuities? Does that seem more tractable?

Not really. The problem is that finding zeros for crossing functions is already complicated, and this adds a new use case with additional logic:

- What if the variable is constant during a certain time. How should that be signaled; it's not an edge during that time - but it's still something that someone might want to know.
- What if the variable is _almost_ constant, but there is numerical noise? This is especially important if the solver uses advanced logic to find multiple crossings for one variable during one step.

**Comment by mestinso** [comment]
> * What if the variable is constant during a certain time. How should that be signaled; it's not an edge during that time - but it's still something that someone might want to know.

Sure, but I don't see the issue, since it's not an edge it should signal false.

> * What if the variable is _almost_ constant, but there is numerical noise? This is especially important if the solver uses advanced logic to find multiple crossings for one variable during one step.

I can see that being an issue, but aren't the solvers already dealing with this in some way already? For example, if I take my continuous edge implementation from the first post and plug in an exponential sin wave, I get the following (in Dymola):
![image](https://user-images.githubusercontent.com/11094799/205663928-021bc20f-1266-4e38-9b0d-fbf21b225e23.png)

The apparent numerical issues that occur at ~25 s I would consider acceptable.

**Comment by HansOlsson** [comment]
> > * What if the variable is _almost_ constant, but there is numerical noise? This is especially important if the solver uses advanced logic to find multiple crossings for one variable during one step.
> 
> I can see that being an issue, but aren't the solvers already dealing with this in some way already? 

Of course, but that is adapted to the current logic - and in particular they are based on Modelica's more robust "x>0" vs "x passes through 0", and that still required additional changes for some operations. I haven't examined whether it is possible to implement this proposal as robustly, and if not considered in detail we will have edge-cases where `x=pre(x)` but `x-pre(x)>0`.

Note that ExpSine is fairly noise-free, whereas anything that is based on solving a non-linear system of equations will be noisier.

**Comment by HansOlsson** [comment]
Apart from the issues with detecting events I also note that `v-pre(v)` is problematic. On one hand one would expect that `pre(v)` was from the latest event involving `v` - or possibly the start of the step, but in general `pre(v)` is defined as the value at the previous event. If there are other sources of events these will differ, and if we e.g. sum up `(v-pre(v))^2` that will be noticeable, which is really problematic.

**Comment by henrikt-ma** [comment]
My expectation would have been that `v - pre(v) = 0` except at discontinuities, and I realize that this shows why approximating `pre(v)` by the value from the last accepted step wouldn't give the expected results.

This makes me believe that that the only reliable definition we could give `pre(v)` is that it is equal to `v` between events, and equal to the value from the previous round of event iteration at events (taking the value right before the event for the first round).  While I think lifting a restriction of `pre` in this way could be elegant, I realize that it wouldn't address the use cases mentioned originally in this issue.

**Comment by mestinso** [comment]
> My expectation would have been that `v - pre(v) = 0` except at discontinuities, and I realize that this shows why approximating `pre(v)` by the value from the last accepted step wouldn't give the expected results.

Agreed, this was basically my line of thinking as well, with the additional conceptual thought that:
when the slope is positive: v-pre(v) is an infinitesimally small positive value
when the slope is negative: v-pre(v) is an infinitesimally small negative value
and when the slope is zero: v-pre(v) is a true zero value
...and ideally (magically?), we could distinguish between these 3 cases to the point that comparisons could be made to detect edges etc.

Again, since the above is numerically/programmatically problematic, I would emphasize: I'm fully ok with throwing out the idea of generalizing `pre` in a way that could be used to detect edges, but then I would just want to see if we can address the `edge` and `change` functions (or new continuous variants) directly instead (in a way that doesn't use `pre` for the continuous input case).

If there is agreement that generalizing `pre` isn't the path forward and that `edge` and `change` functions could be addressed instead, does it make sense to close this Issue and open another separate one to focus the discussion?

**Comment by henrikt-ma** [comment]
I'm afraid that I can't see clearly enough how this could be achieved, so don't think I can be of any more help until there's something more tangible to think through.

**Comment by eshmoylova** [comment]
Why not use `der` to detect changes in slope for continuous variables? The reason we need `pre` and difference equations for discrete variables is because derivative is not defined. But we do have derivative for continuous variables. In your proposal, did you intend `pre` to give value at a previous event or at a previous iteration step?

**Comment by mestinso** [comment]
> Why not use `der` to detect changes in slope for continuous variables? The reason we need `pre` and difference equations for discrete variables is because derivative is not defined. But we do have derivative for continuous variables. In your proposal, did you intend `pre` to give value at a previous event or at a previous iteration step?

In addition to what @HansOlsson said, I would emphasize that, from my point of view, the Modelica language (as is) is sufficient for detecting rising edge, falling edge, and change for situations without discontinuities (by using der as shown in the ContinuousEdge function in the first post). The problem is situations with discontinuities, like step functions and sawtooth functions. 

As mentioned, this functionality works fine in Simulink. So, in that environment, this is what we get for a sawtooth paired with a rising edge detector:
![image](https://user-images.githubusercontent.com/11094799/207391186-cd4c2ebc-7fba-4087-96b2-659716eb8b8e.png)

And this is what we get for a step function paired with a rising edge detector:
![image](https://user-images.githubusercontent.com/11094799/207391561-ea22679a-dc54-4fa3-b90a-518e1cade447.png)

The key feature/language addition needed is the special handling through the step changes to make this work.
